src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs

```diff
@@ -42,9 +43,35 @@ protected override void InitializeWorker(AnalysisContext context)
 
         private void AnalyzeDiscardDesignation(SyntaxNodeAnalysisContext context)
         {
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+
             var discard = (DiscardDesignationSyntax)context.Node;
-            if (discard.Parent is DeclarationPatternSyntax)
+
+            if (discard.Parent is DeclarationPatternSyntax declarationPattern)
             {
+                // Don't perform semantic checks if we are in a 'simple' pattern like `x is A _`.
+                // Since single identifier in cases like `x is A` binds stronger to type name for back compat reasons, we can safely remove discard anyway.
+                if (declarationPattern.Parent is not IsPatternExpressionSyntax)
+                {
+                    var typeSyntax = declarationPattern.Type;
+
+                    if (typeSyntax is QualifiedNameSyntax qualifiedName)
+                        typeSyntax = qualifiedName.GetLeftmostName();
+
+                    if (typeSyntax is IdentifierNameSyntax identifierName &&
+                        identifierName.GetAncestor<TypeDeclarationSyntax>() is { } containingTypeSyntax)
+                    {
+                        var typeSymbol = semanticModel.GetDeclaredSymbol(containingTypeSyntax, cancellationToken);
```

Comment:
```suggestion
                        var typeSymbol = semanticModel.GetRequiredDeclaredSymbol(containingTypeSyntax, cancellationToken);
```

this can't fail.

Reply:
`GetRequiredDeclaredSymbol` returns `ISymbol`, so if we apply it, explicit cast will also be needed. Is it worth being explicit about the fact of always succeeding semantic search, but downgrade readability by having the cast?

Reply:
you can add an overload of GetRequiredDeclardSymbol with the same strong typing as GetDeclaredSymbol.